cabal-install/Distribution/Client/ProjectPlanning.hs

```diff
@@ -294,48 +294,61 @@ sanityCheckElaboratedPackage ElaboratedConfiguredPackage{..}
 rebuildProjectConfig :: Verbosity
                      -> DistDirLayout
                      -> ProjectConfig
-                     -> IO (ProjectConfig,
-                            [PackageSpecifier UnresolvedSourcePackage])
-rebuildProjectConfig verbosity
-                     distDirLayout@DistDirLayout {
-                       distProjectRootDirectory,
-                       distDirectory,
-                       distProjectCacheFile,
-                       distProjectCacheDirectory
-                     }
-                     cliConfig = do
+                     -> IO
+                          ( ProjectConfig
+                          , [PackageSpecifier UnresolvedSourcePackage]
+                          )
+rebuildProjectConfig verbosity distDirLayout cliConfig = do
 
     (projectConfig, localPackages) <-
-      runRebuild distProjectRootDirectory $
-      rerunIfChanged verbosity fileMonitorProjectConfig () $ do
-
-        projectConfig <- phaseReadProjectConfig
-        localPackages <- phaseReadLocalPackages projectConfig
-        return (projectConfig, localPackages)
+      runRebuild distProjectRootDirectory
+      $ rerunIfChanged verbosity
+                       fileMonitorProjectConfig
+                       (projectConfigConfigFile, distProjectFile "")
+      $ do
+          liftIO $ info verbosity "Project settings changed, reconfiguring..."
+          projectConfig <- phaseReadProjectConfig
+          localPackages <- phaseReadLocalPackages projectConfig
+          return (projectConfig, localPackages)
+
+    info verbosity
+      $ unlines
+      $ ("this build was affected by the following (project) config files:" :)
+      $ [ path
+        | Explicit path <- Set.toList $ projectConfigProvenance projectConfig
+        ]
 
     return (projectConfig <> cliConfig, localPackages)
 
   where
-    ProjectConfigShared {
-      projectConfigConfigFile
-    } = projectConfigShared cliConfig
-
-    fileMonitorProjectConfig = newFileMonitor (distProjectCacheFile "config")
+    DistDirLayout { distProjectRootDirectory
+                  , distDirectory
+                  , distProjectCacheFile
+                  , distProjectCacheDirectory
+                  , distProjectFile
+                  }
+      = distDirLayout
+    ProjectConfigShared { projectConfigConfigFile } =
+      projectConfigShared cliConfig
+
+    fileMonitorProjectConfig =
+      newFileMonitor (distProjectCacheFile "config") :: FileMonitor
+          (Flag FilePath, FilePath)
```

Comment:
we should not pass a `Flag FilePth` into the monitor.  I suggest you use `getConfigFilePath` to resolve to properly resolve the config file path.

Reply:
Ah, good I added that type signature to counteract the `NoMonoLocalBinds` then :D